_posts/2020-06-03-using-Torch-with-Flux-2.md

```diff
@@ -0,0 +1,116 @@
+
+In that spirit, we introduce [Torch.jl](https://github.com/FluxML/Torch.jl), a package that wraps the optimised kernels written for torch and PyTorch, and makes them available to use through Julia. We expect to do this to gain coverage over our existing GPU stack, and bring us plenty of performance improvements on GPUs along the way. As an example we take a popular object detection model - ResNet50 - and compare it with our existing tooling.
+
+![img](https://user-images.githubusercontent.com/20618666/79492160-e60a5600-801f-11ea-9d7f-dd04af243e44.png)
+*Image by github user https://github.com/gartangh*
+
+Not too shabby, we see about 2 orders of magnitude worth of improvement in runtime over the native kernels (all packages make use of the optimised cudnn kernels where possible). We however, see some areas of pending improvement. One of them being memory management. This is one of the areas which we intend to improve upon, a source of majority of overhead. 
+
+## Usage
+
+Adding the package using the Pkg, Julia's package manager is easy. However, note that at the time of writing, Torch.jl assumes the presence of a CUDA enabled GPU on the device it's being added to, and assumes linux as the OS.
+
+```julia
+# Type `]` to enter Pkg mode in the Julia REPL.
+pkg> add Torch
+[...] # Note that this downloads the Torch artifact, which is quite large
+
+julia> using Torch
+```
+
+Once it has been installed, we can talk about the different functions that Torch.jl makes available.
+
+In addition, we expect to make as few but relevant assumptions about the kind of models that the ML community has been developing, and how we see their use grow in more fields than ever before, which is why we want these `Tensor`s to mimic Julia arrays closely
+
+Torch.jl provides the `Tensor` type which closely follows the semantics of a regular Julia array, albeit being managed by torch. One can create a tensor with an API similar to `rand` or `zeros` etc.
+
+```julia
+julia> z = Tensor(3,3)
+3×3 Tensor{Float32,2} :
+ 0.0  0.0  0.0
+ 0.0  0.0  0.0
+ 0.0  0.0  0.0
+```
+
+To control the device the tensor is loaded on (the default being on CPU), we use the `dev` keyword, available in most functions.
+
+```julia
+julia> z = Tensor(3,3, dev = 0)
+3×3 Tensor{Float32,2} :
+ 0.0  0.0  0.0
+ 0.0  0.0  0.0
+ 0.0  0.0  0.0
+```
+
+Note that setting `dev` to `-1` implies the CPU, and `[0,...]` represent the id of the GPU we intend to load the tensor on. The default GPU is assumed to be `0`, for functions revelant to moving these tensors around. Torch.jl also emits the `torch` function which behaves like the `gpu` function already in Flux, moving over structs to Torch instead of CUDA.
+
+```julia
+julia> using Flux, Metalhead, Torch
+
+julia> using Torch: torch
+
+julia> resnet = ResNet() # from Metalhead
+ResNet()
+
+julia> tresnet = resnet |> torch
+ResNet()
+```
```

Comment:
Mention VGG, GoogleNet, DenseNet etc. are also available in MetalHead and link to the docs - say they will all benefit (if that is indeed the case).

Reply:
Done